### PR TITLE
Don't send RST_STREAM when not needed

### LINKED ISSUE #433: Avoid unnecessary `RST_STREAM` frames in HTTP/2 implementation
When a stream is locally closed and a EndStream flag is received, Mint is sending RST_STREAM with NO_ERROR.
This is unnecessary as a stream can be deemed closed by both sides at this point, and this frame is just ignored.

### PR DESCRIPTION
**Pull request by v0idpwn**
When the stream state is local_half_closed and an end_stream flag is received, the stream can be terminated normally without RST_STREAM.

Closes #433.

I still didn't work on integration tests, and while current tests kind of cover this feature, I'll add a specific test for it.

### CODE REVIEW

**Comment by v0idpwn** [comment]
This is a WIP, I'm opening for reference. I need to double check some corner cases and maybe my implementation can be simplified.

File: lib/mint/http2.ex [hunk 1]
```diff
@@ -2094,18 +2094,27 @@ defmodule Mint.HTTP2 do
...
+      cond do
+        # If the stream is ended on both sides, it is already deemed closed and
+        # there's no need to send a RST_STREAM frame
+        reason == :remote_end_stream and stream.state == :half_closed_local ->
+          conn
+
+        # We send a RST_STREAM with the given error code so that we move the
+        # stream to the :closed state (that is, we remove it).
+        open?(conn) ->
+          error_code = if reason == :remote_end_stream, do: :no_error, else: reason
```
Comment (v0idpwn) [inline lib/mint/http2.ex:2111]: I don't think this line is ever used, but it doesn't feel right to remove it (?)

**Review by v0idpwn — COMMENTED** [review]
I'm not 100% happy with the implementation because I think it's a bit dirty, but it solves the problem and is pretty small, so I'm opening the PR.

**Comment by v0idpwn** [comment]
Also it looks like CI is failing for totally unrelated reasons™️

**Comment by whatyouhide** [comment]
> Also it looks like CI is failing for totally unrelated reasons™️

Yes Totally Unrelated™, no worries

> I'm not 100% happy with the implementation because I think it's a bit dirty, but it solves the problem and is pretty small, so I'm opening the PR.

@v0idpwn is this ready to review?

**Comment by v0idpwn** [comment]
Yes!

**Comment by whatyouhide** [comment]
Woot woot!
